urdf/src/model.cpp

```diff
@@ -108,21 +110,22 @@ bool Model::initParamWithNodeHandle(const std::string& param, const ros::NodeHan
   return Model::initString(xml_string);
 }
 
-bool Model::initXml(TiXmlDocument *xml_doc)
+bool Model::initXml(XMLDocument *xml_doc)
 {
   if (!xml_doc)
   {
     ROS_ERROR("Could not parse the xml document");
     return false;
   }
 
-  std::stringstream ss;
-  ss << *xml_doc;
+  XMLPrinter printer;
+  xml_doc->Print(&printer);
+  std::string str(printer.CStr());
 
-  return Model::initString(ss.str());
+  return Model::initString(str);
```

Comment:
@clalancette It appears the `initXml(<TinyXML type>)` APIs are only convenience for parsing from a string. How do you feel about eliminating `TinyXML`  altogether by removing these API from Melodic?

Reply:
Ha, I didn't realize that these were just printing to a string so that we could go back in via the `std::string` version of the API.  That's kind of silly.

However, my opinion on removing it is a little more nuanced.  I'm somewhat concerned that we will litter the above XML->String code over a lot of different packages, thereby introducing the possibility of bugs, and making it harder to fix in the future.  Before doing that, I think it would be worthwhile to take a look at how many downstream users use this interface vs. the string one.  If that number is fairly low, I'd be more in favor of removing these APIs.

Reply:
I ended up doing a bit of digging here.

First, I ran a prerelease script on indigo down to 4 levels for urdf.  This ended up pulling in 677 packages.  From there, I ended up grepping for `initXml` across all of the sources, which landed me a much smaller number of packages that actually call `initXml`.  Here's the list:

* collada_urdf
* kdl_parser
* moveit_ros_planning
* open_controllers_interface
* pr2_arm_kinematics
* pr2_arm_move_ik
* pr2_controller_manager
* pr2_ethercat
* pr2_gazebo plugins
* pr2_mechanism_controllers
* pr2_mechanism_model
* pr2_moveit_plugins
* ros_ethercat_model
* rviz
* sr_mechanism_controllers

I didn't exhaustively look at each instance of the `initXml` code in those packages (this is feasible, but tedious), but it looks like the majority of them use the `Model::initXml(TiXmlElement *)` variant of the API.  This leads me to two things:

1.  I think we need to keep the version where we expose the tinyxml(2) API, since most users expect that.
1.  This transition is going to be very, very painful for downstream users.  We have control of a couple of the users (`collada_urdf` and `kdl_parser`), but not most of them.  And this doesn't even address the closed-source users.  I'm starting to think that we may really have to add a tinyxml backwards-compatibility layer, because otherwise we are just going to break the world.